File: kernel/src/process/elf.rs
use elf_rs::{Elf, ElfFile, ProgramHeaderFlags, ProgramType};
use log::info;
use x86_64::{
    structures::paging::{FrameAllocator, Mapper, Page, PageSize, PageTableFlags, Size4KiB},
    VirtAddr,
};

use crate::{
    kernel::paging::{KernelPaging, UserspacePaging},
    process::MappedPage,
};

pub fn load<T: AsRef<str>>(
    process_paging: &mut UserspacePaging,
    kernel_paging: &mut KernelPaging,
    bytes: &[u8],
    args: &[T],
) -> (VirtAddr, VirtAddr, alloc::vec::Vec<MappedPage>) {
    let elf = Elf::from_bytes(bytes).expect("Failed to parse ELF!");
    match elf.elf_header().elftype() {
        elf_rs::ElfType::ET_EXEC => {}
        e => panic!("Unsupported ELF type: {:?}", e),
    }

    let mut pages = alloc::vec::Vec::with_capacity(10);

    // map pages
    for ph in elf
        .program_header_iter()
        .filter(|ph| ph.ph_type() == ProgramType::LOAD)
    {
        let start_page = Page::containing_address(VirtAddr::new(ph.vaddr()));
        let end_page = Page::containing_address(VirtAddr::new(ph.vaddr() + ph.memsz()));

        let mut flags = PageTableFlags::PRESENT | PageTableFlags::USER_ACCESSIBLE;
        if !ph.flags().contains(ProgramHeaderFlags::EXECUTE) {
            flags |= PageTableFlags::NO_EXECUTE;
        }
        if ph.flags().contains(ProgramHeaderFlags::WRITE) {
            flags |= PageTableFlags::WRITABLE;
        }

        info!(
            "mapping code pages [{:#0x} - {:#0x}] with flags {:?}",
            start_page.start_address(),
            end_page.start_address(),
            flags
        );

        // map page for copying program
        for page in Page::range_inclusive(start_page, end_page) {
            pages.push(MappedPage {
                name: "elf",
                page,
                flags,
            });

            let frame = kernel_paging
                .allocate_frame()
                .expect("Failed to allocate frame!");

            unsafe {
                process_paging
                    .page_table
                    .map_to(
                        page,
                        frame,
                        PageTableFlags::PRESENT | PageTableFlags::WRITABLE,
                        &mut *kernel_paging,
                    )
                    .unwrap_or_else(|e| panic!("Failed to map page {:#?}: {:?}", page, e))
                    .flush();
            }
        }

        unsafe {
            process_paging.load();

            // copy data
            core::ptr::copy::<u8>(
                bytes.as_ptr().add(ph.offset() as usize),
                ph.vaddr() as *mut u8,
                ph.filesz() as usize,
            );

            // zero out the rest of the page
            let remaining_size = ph.memsz() as usize - ph.filesz() as usize;
            core::ptr::write_bytes((ph.vaddr() + ph.filesz()) as *mut u8, 0, remaining_size);

            kernel_paging.load();
        }

        // remap the page to the correct flags
        for page in Page::range_inclusive(start_page, end_page) {
            unsafe {
                let (frame, flush) = process_paging
                    .page_table
                    .unmap(page)
                    .expect("Failed to unmap page!");
                flush.flush();

                process_paging
                    .page_table
                    .map_to(page, frame, flags, &mut *kernel_paging)
                    .expect("Failed to remap page!")
                    .flush();
            }
        }
    }

    // create stack
    let stack_size_pages = 100;
    let stack_address = VirtAddr::new(0x0000_1000_0000_0000);
    let flags = PageTableFlags::PRESENT
        | PageTableFlags::WRITABLE
        | PageTableFlags::USER_ACCESSIBLE
        | PageTableFlags::NO_EXECUTE;
    info!(
        "mapping stack pages [{:#0x} - {:#0x}] with flags {:?}",
        stack_address,
        stack_address + stack_size_pages * Size4KiB::SIZE,
        flags
    );
    for i in 0..stack_size_pages {
        let page = Page::containing_address(stack_address + i * Size4KiB::SIZE);
        let frame = kernel_paging
            .allocate_frame()
            .expect("Failed to allocate frame!");

        unsafe {
            process_paging
                .page_table
                .map_to(page, frame, flags, &mut *kernel_paging)
                .expect("Failed to map page!")
                .flush();
        }

        pages.push(MappedPage {
            name: "stack",
            page,
            flags,
        });
    }

    let stack_top = stack_address + stack_size_pages * Size4KiB::SIZE;

    // prepare arguments on the stack

    let stack_top_frame = process_paging
        .page_table
        .translate_page(Page::<Size4KiB>::containing_address(stack_top - 1))
        .expect("Failed to translate stack top page");

    log::debug!("preparing arguments on the stack at {stack_top:#x} (frame: {stack_top_frame:x?})");

    {
        let l4_index = Page::<Size4KiB>::containing_address(stack_top - 1)
            .start_address()
            .p4_index();
        let kernel_l4_ptr = &kernel_paging.page_table().level_4_table()[l4_index];
        let process_l4_ptr = &process_paging.page_table.level_4_table()[l4_index];
        log::log!(
            log::Level::Debug,
            "l4_index: {l4_index:?}, kernel L4 entry: {:#x} ({:x?}), process L4 entry: {:#x} ({:x?})",
            kernel_l4_ptr as *const _ as u64,
            kernel_l4_ptr,
            process_l4_ptr as *const _ as u64,
            process_l4_ptr
        );
    }

    unsafe {
        kernel_paging
            .map_to(
                Page::<Size4KiB>::containing_address(stack_top - 1),
                stack_top_frame,
                PageTableFlags::PRESENT | PageTableFlags::WRITABLE,
                &mut crate::kernel::paging::UseKernelFrameAllocator,
            )
            .expect("failed to map stack top page")
            .flush();
    }

    // copy all arg string contents to the stack
    let str_area = stack_top - args.iter().map(|s| s.as_ref().len()).sum::<usize>() as u64;
    log::debug!("arg string area starts at {str_area:#x}");
    let mut offset = 0;
    for arg in args {
        let ptr = unsafe { str_area.as_mut_ptr::<u8>().add(offset) };
        unsafe { core::ptr::copy_nonoverlapping(arg.as_ref().as_ptr(), ptr, arg.as_ref().len()) };

        offset += arg.as_ref().len();
    }

    // create string structs for each argument
    let arg_str_ptr = (str_area - (args.len() * size_of::<crate::types::string_const_t>()) as u64)
        .align_down(align_of::<crate::types::string_const_t>() as u64);
    log::debug!("arg string structs starts at {arg_str_ptr:#x}");
    let mut offset = 0;
    for (i, arg) in args.iter().enumerate() {
        let arg_ptr = unsafe {
            arg_str_ptr
                .as_mut_ptr::<crate::types::string_const_t>()
                .add(i)
        };
        unsafe {
            *arg_ptr = crate::types::string_const_t {
                ptr: str_area.as_ptr::<i8>().add(offset),
                len: arg.as_ref().len() as u32,
            };
        }
        offset += arg.as_ref().len();
    }

    // create entry struct
    let entry_struct_ptr = (arg_str_ptr - size_of::<crate::types::entry_t>() as u64)
        .align_down(align_of::<crate::types::entry_t>() as u64);
    log::debug!("entry pointer starts at {entry_struct_ptr:#x}");
    let entry_args_ptr = entry_struct_ptr.as_mut_ptr::<crate::types::entry_t>();

    unsafe {
        *entry_args_ptr = crate::types::entry_t {
            argc: args.len() as u32,
            argv: arg_str_ptr.as_mut_ptr::<crate::types::string_const_t>(),
        };
    }

    // create pointer to the entry struct
    let entry_ptr = entry_struct_ptr - size_of::<u64>() as u64;
    log::debug!("entry pointer is at {entry_ptr:#x}");
    unsafe {
        *(entry_ptr.as_mut_ptr::<u64>()) = entry_args_ptr as u64;
    }

    kernel_paging
        .unmap(Page::<Size4KiB>::containing_address(stack_top - 1))
        .expect("Failed to unmap stack top page")
        .1
        .flush();

    (VirtAddr::new(elf.entry_point()), entry_ptr, pages)
}

File: kernel/src/process/elf.zig
const std = @import("std");

const paging = @import("../paging.zig");

const Header = packed struct {
    type: enum(u16) {
        ET_NONE = 0,
        ET_REL = 1,
        ET_EXEC = 2,
        ET_DYN = 3,
        ET_CORE = 4,
        _,
    },
    machine: enum(u16) {
        EM_X86_64 = 62,
        _,
    },
    version: u32,
    entry: u64,
    phoff: u64,
    shoff: u64,
    flags: u32,
    ehsize: u16,
    phentsize: u16,
    phnum: u16,
    shentsize: u16,
    shnum: u16,
    shstrndx: u16,
};

const ProgramHeader = packed struct(u448) {
    type: enum(u32) {
        PT_NULL = 0,
        PT_LOAD = 1,
        PT_DYNAMIC = 2,
        PT_INTERP = 3,
        PT_NOTE = 4,
        PT_SHLIB = 5,
        PT_PHDR = 6,
        _,
    },
    flags: u32,
    offset: u64,
    vaddr: u64,
    paddr: u64,
    filesz: u64,
    memsz: u64,
    @"align": u64,
};

pub const MappedPage = struct {
    virtualAddress: u64,
    size: paging.FrameSize,
};

pub const Elf = struct {
    bytes: []const u8,
    entry: u64,

    fn ident(self: Elf) []const u8 {
        return self.bytes[0..16];
    }

    fn header(self: Elf) Header {
        return std.mem.bytesToValue(Header, self.bytes[16 .. 16 + @sizeOf(Header)]);
    }

    pub fn load(
        bytes: []const u8,
        pagetable: *paging.OffsetPageTable,
        frameAllocator: *paging.FrameAllocator,
        mappedPages: *std.ArrayList(MappedPage),
    ) !Elf {
        pagetable.load();

        const self = Elf{ .bytes = bytes, .entry = 0 };

        // confirm ELF magic number
        if (!std.mem.startsWith(u8, self.ident(), "\x7fELF")) {
            return error.InvalidElf;
        }

        const h = self.header();
        if (h.type != .ET_EXEC) {
            return error.InvalidElf;
        }
        if (h.machine != .EM_X86_64) {
            return error.InvalidElfMachine;
        }
        if (h.version != 1) {
            return error.InvalidElfVersion;
        }

        std.log.debug("ELF header: type: {d}, machine: {d}, version: {d}, entry: 0x{x}, phoff: 0x{x}, shoff: 0x{x}, flags: {d}, ehsize: {d}, phentsize: {d}, phnum: {d}", .{
            h.type, h.machine, h.version, h.entry, h.phoff, h.shoff, h.flags, h.ehsize, h.phentsize, h.phnum,
        });

        for (0..h.phnum) |i| {
            const phdr = @as(*align(1) const ProgramHeader, @ptrCast(&self.bytes[h.phoff + i * h.phentsize]));

            switch (phdr.type) {
                .PT_LOAD => {
                    const startPage = phdr.vaddr & ~@as(u64, 0xFFF);
                    const endPage = (phdr.vaddr + phdr.memsz) & ~@as(u64, 0xFFF);

                    var page = startPage;
                    while (page <= endPage) : (page += 0x1000) {
                        // skip if the page is already mapped
                        if (pagetable.translate(page) != null) {
                            std.log.debug("skipping already mapped page 0x{x}", .{page});
                            continue;
                        }

                        const frame = try frameAllocator.allocateFrame(.@"4KiB");
                        try pagetable.map(
                            frameAllocator,
                            page,
                            frame,
                            .@"4KiB",
                            .{
                                .userAccessible = true,
                                //.writable = (phdr.flags & 0x2) != 0,
                                .writable = true,
                                .noExecute = (phdr.flags & 0x1) == 0,
                            },
                        );

                        // zero out the page
                        const memory: []u8 = @as([*]u8, @ptrFromInt(page))[0..0x1000];
                        @memset(memory, 0);

                        try mappedPages.append(.{ .virtualAddress = page, .size = .@"4KiB" });
                    }

                    // copy the segment data into the mapped memory
                    const memory: []u8 = @as([*]u8, @ptrFromInt(phdr.vaddr))[0..phdr.memsz];
                    const data: []const u8 = self.bytes[phdr.offset .. phdr.offset + phdr.filesz];
                    std.mem.copyForwards(u8, memory, data);
                },
                else => {
                    std.log.warn("ignoring segment {d}: type: {d}", .{ i, phdr.type });
                },
            }
        }

        return Elf{
            .bytes = bytes,
            .entry = h.entry,
        };
    }
};
